### DISCUSSION TITLE
Enabling -M argument < 10% -- probably needs percentage reader in transient or allow leading zeros in transient--read-number-N

### DISCUSSION THREAD

**Discussion by joodie** [magit/magit#4708] [Ideas]
Hi everybody! First time using the discussion feature here. Please let me know if there's anything I can do to improve the discussion.

Anyway, today I was trying to run the equivalent of `git diff -w -M01 rev1..rev2` - I was working on a changeset that moved and changed a few files and then I had to deal with concurrent changes, so the reason for the -M01 setting is to lower the "files moved" detection in git to 1% match. In my case a 10% match is too high (probably because of whitespace changes on many lines)

If I try to enter this option in the magit diff popup, I cannot enter a number with leading zeros, because it uses `transient-read-number-N+` which expects a natural number > 1 and explicitly disallows leading zeros.

The problem is that the argument to `-M` is not a natural number but a percentage or fraction;

   - Percentages can be input in the form of a number directly followed by a percentage sign (%).
   - Fractions can be input as one or more digits without any following sign, and are interpreted as "0." followed by the given digits; this means that "1" gets parsed as "0.1" which is 10%, and "01" is parsed as "0.01" which is 1%.

Neither of the options match transient-read-number-N+ if you try to enter a setting less than 10%

My thinking is that to solve this a `transient--read-percentage` or `transient--read-fraction` function will have to be added to allow for this kind of input in a clean way. Alternatively `transient--read-number-N` could be expanded to allow leading zeros, but I'm not sure of the impact of that (parsing numbers as octal maybe?)

I'm willing to provide patches if there is consensus on the right way to implement this.

**Comment by tarsius** [comment]
I believe to remember that I actively decided against supporting all valid values because of the added complexity and opportunity cost. Also I feel that a similarity smaller than the already very small 10% is not very useful, but I guess you have a use-case for that.

While the current implementation doesn't allow all valid forms, it disallows invalid forms. A new implementation that supports all valid values would have to keep the ability to detect invalid input, as it is being typed. It also shouldn't be large.

**Reply by tarsius** [reply]
Implemented in bed0bb233e61be8a8378ed7ad4dcc78962bc665b.
